File: Quiz_12/VaniAgrawal_590015846/Quiz_12_Debugging_Question_2.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def range_sum_bst(root, L, R):
    if not root:
        return 0
    if root.val < L:
        return range_sum_bst(root.right, L, R)
    if root.val > R:
        return range_sum_bst(root.left, L, R)
    return root.val + range_sum_bst(root.left, L, R) + range_sum_bst(root.right, L, R)

# Task 1: Identify the specific flaw in the validation logic
# The original code does not skip unnecessary branches and does not efficiently prune the search space according to the BST property.

# Task 2: Why this causes incorrect behavior
# The original logic traverses all nodes regardless of their relevance to the range, causing inefficiency and potentially incorrect sums if irrelevant nodes are included.

# Correct implementation ensures pruning by comparing node values with L and R and only traversing relevant subtrees.

root = TreeNode(10)
root.left = TreeNode(5)
root.right = TreeNode(15)
root.left.left = TreeNode(3)
root.left.right = TreeNode(7)
root.right.right = TreeNode(18)

L = 7
R = 15

result = range_sum_bst(root, L, R)
print(f"Range sum is: {result}")

File: Quiz_12/AKSHCHAUHAN_500120187/Quiz_12_Debugging_Question_2.c
#include <stdio.h>
#include <stdlib.h>

struct TreeNode {
    int val;
    struct TreeNode* left;
    struct TreeNode* right;
};

struct TreeNode* newNode(int val) {
    struct TreeNode* node = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    node->val = val;
    node->left = node->right = NULL;
    return node;
}

int rangeSumBST(struct TreeNode* root, int L, int R) {
    if (root == NULL) 
        return 0;
    if (root->val < L) 
        return rangeSumBST(root->right, L, R);
    if (root->val > R) 
        return rangeSumBST(root->left, L, R);
    return root->val 
         + rangeSumBST(root->left, L, R) 
         + rangeSumBST(root->right, L, R);
}

int main() {
    struct TreeNode* root = newNode(10);
    root->left = newNode(5);
    root->right = newNode(15);
    root->left->left = newNode(3);
    root->left->right = newNode(7);
    root->right->left = newNode(13);
    root->right->right = newNode(18);
    root->left->left->left = newNode(1);
    root->left->right->left = newNode(6);
    root->right->right->right = newNode(20);

    int L = 7, R = 15;
    int result = rangeSumBST(root, L, R);

    printf("Range sum [%d, %d]: %d\n", L, R, result);

    return 0;
}